Improve simplification of hyperbolic functions

ylemkimon opened the pull request:
- Fixed incorrect hyper to trig conversion in hyper_as_trig(_osbornei) when constant is intoduced during simplification: Fixes #11685.
- Fixed simplifying expressions with sech or csch raising NotImplementedError.
- Improved tan/cot _peeloff_pi logic, as m is guaranteed to be Rational multiple of pi/2.
- Added _peeloff_ipi and simplification in hyperbolic functions, which makes Rational pi coefficient of imaginary part to be in [0, pi/2), same behavior as trigonometric functions.

smichr commented on sympy/simplify/fu.py:
```diff
@@ -2103,6 +2113,9 @@ def hyper_as_trig(rv):
         t, f = hyper_as_trig(expr)
         expr == f(t)
 
+    Be careful in simplifying returned expression as some trigonometric
```
This makes me nervous -- can you give an example?

smichr replied:
I see #11685, but didn't you fix TR10i so now this warning wouldn't apply?

ylemkimon replied:
@smichr As TR10i fixes this by undocumented private _hyper flag, it is not directly applicable. Also, I think it is possible that user or we may implement new trigonometric simplification or simplify the expression themselves.

smichr replied:
Aren't all manipulations of the transformed hyperbolic functions valid? I don't know that much about it so forgive the ignorance.

ylemkimon replied:
@smichr When transforming hyperbolic functions to trigonometric functions and vice versa, `I` is multiplied to the arguments. The `hyper_as_trig()` does this by multiplying dummy variable when transforming hyperbolic functions to trigonometric functions, and substituting dummy variable with -1 when transforming back, assuming argument is multiplied by the dummy variable. However, this fails when constant is introduced to the arguments, which should be multiplied by I when transforming back.

smichr replied:
Ahh...that makes sense. Can't we fix this by making sure there is no numerical coefficient on the result when we convert back to hyperbolics? e.g. if we had j*f(x) and end up with c + j*g(x) after simplification, then can't we replace c with c*j? But can you give a specific example of such a c being introduced?

ylemkimon replied:
@smichr `I*sqrt(3)*cosh(x)/2 + sinh(x)/2` -_osborne-> `I*sqrt(3)*cos(d*x)/2 + I*sin(d*x)/2` -TR10i-> `I*sin(d*x + pi/3)` -_osbornei-> `sinh(x + pi/3)` where `I*sqrt(3)*cosh(x)/2 + sinh(x)/2 == sinh(x + pi/3*I)`.

ylemkimon replied:
@smichr Maybe like below?:
```diff
diff --git a/sympy/simplify/fu.py b/sympy/simplify/fu.py
index d07bffbfe..170397520 100644
--- a/sympy/simplify/fu.py
+++ b/sympy/simplify/fu.py
@@ -2084,7 +2084,8 @@ def _osbornei(e, d):
     def f(rv):
         if not isinstance(rv, TrigonometricFunction):
             return rv
-        a = rv.args[0].xreplace({d: S.One})
+        const, a = rv.args[0].as_independent(d)
+        a = a.xreplace({d: S.One}) + const*I
         if rv.func is sin:
             return sinh(a)/I
```

smichr replied:
Yes -- but use flag `as_Add=True` with `as_independent` to ensure that you are getting an additive constant.

Also, I am having a hard time following the example that you gave. Could you show me an explicit interactive session where you end up with that additive constant?

ylemkimon committed:
Improve tan/cot _peeloff_pi logic

Improved tan/cot _peeloff_pi logic, as m is guaranteed to be Rational
and multiple of pi/2.

ylemkimon committed:
Add _peeloff_ipi in hyperbolic functions

Added _peeloff_ipi and simplification by it in hyperbolic functions,
which makes Rational pi coefficient of imaginary part to be in [0,
pi/2), like trigonometric functions.

ylemkimon committed:
Fix incorrect _osbornei when constant is introduced

Fixed incorrect hyper to trig conversion in hyper_as_trig(_osbornei)
when constant is intoduced during simplification. See sympy/sympy#13259.

ylemkimon commented:
@smichr
Changed 😃 

```
In [1]: from sympy.utilities.randtest import verify_numerically

In [2]: from sympy.simplify.fu import TR10i, hyper_as_trig

In [3]: a = I*sqrt(3)*cosh(x)/2 + sinh(x)/2

In [4]: b = simplify(a)

In [5]: b
Out[5]:
    /    pi\
sinh|x + --|
    \    3 /

In [6]: verify_numerically(a, b)
Out[6]: False

In [7]: t, f = hyper_as_trig(a)

In [8]: t
Out[8]:
                       ___
I*sin(Dummy_149*x)   \/ 3 *I*cos(Dummy_149*x)
------------------ + ------------------------
        2                       2

In [9]: TR10i(t)
Out[9]:
     /              pi\
I*sin|Dummy_149*x + --|
     \              3 /

In [10]: f(_)
Out[10]:
    /    pi\
sinh|x + --|
    \    3 /
```
Also, please see [added tests](https://github.com/sympy/sympy/pull/13259/files#diff-da93f92a6845d41a353f1a3a7736c7cf) and #11685.

smichr commented:
Nice work! It's in.
